libsession/src/main/java/org/session/libsession/messaging/sending_receiving/MessageReceiverHandler.kt

```diff
@@ -188,173 +200,293 @@ fun MessageReceiver.handleVisibleMessage(message: VisibleMessage, proto: SignalS
     val userPublicKey = storage.getUserPublicKey()!!
-    val groupPublicKey = kind.groupPublicKey.toHexString()
-    val groupID = GroupUtil.getEncodedClosedGroupID(GroupUtil.getEncodedClosedGroupID(Hex.fromStringCondensed(groupPublicKey)).toByteArray()) //double encoded
-    val group = storage.getGroup(groupID)
-    if (group == null) {
-        Log.d("Loki", "Ignoring closed group sender key request for nonexistent group.")
+    val userKeyPair = storage.getUserX25519KeyPair()
+    // Unwrap the message
+    val groupID = GroupUtil.doubleEncodeGroupID(groupPublicKey)
+    val group = storage.getGroup(groupID) ?: run {
+        Log.d("Loki", "Ignoring closed group info message for nonexistent group.")
+        return
+    }
+    if (!group.admins.map { it.toString() }.contains(senderPublicKey)) {
+        android.util.Log.d("Loki", "Ignoring closed group encryption key pair from non-admin.")
+        return
+    }
+    // Find our wrapper and decrypt it if possible
+    val wrapper = kind.wrappers.firstOrNull { it.publicKey!!.toByteArray().toHexString() == userPublicKey } ?: return
+    val encryptedKeyPair = wrapper.encryptedKeyPair!!.toByteArray()
+    val plaintext = MessageReceiverDecryption.decryptWithSessionProtocol(encryptedKeyPair, userKeyPair).first
+    // Parse it
+    val proto = SignalServiceProtos.KeyPair.parseFrom(plaintext)
+    val keyPair = ECKeyPair(DjbECPublicKey(proto.publicKey.toByteArray().removing05PrefixIfNeeded()), DjbECPrivateKey(proto.privateKey.toByteArray()))
+    // Store it
+    storage.addClosedGroupEncryptionKeyPair(keyPair, groupPublicKey)
+    Log.d("Loki", "Received a new closed group encryption key pair")
+}
+
+private fun MessageReceiver.handleClosedGroupNameChanged(message: ClosedGroupControlMessage) {
+    val context = MessagingConfiguration.shared.context
+    val storage = MessagingConfiguration.shared.storage
+    val senderPublicKey = message.sender ?: return
+    val kind = message.kind!! as? ClosedGroupControlMessage.Kind.NameChange ?: return
+    val groupPublicKey = message.groupPublicKey ?: return
+    // Check that the sender is a member of the group (before the update)
+    val groupID = GroupUtil.doubleEncodeGroupID(groupPublicKey)
+    val group = storage.getGroup(groupID) ?: run {
+        Log.d("Loki", "Ignoring closed group info message for nonexistent group.")
+        return
+    }
+    // Check common group update logic
+    if (!isValidGroupUpdate(group, message.sentTimestamp!!, senderPublicKey)) {
+        return
+    }
+    val members = group.members.map { it.serialize() }
+    val admins = group.admins.map { it.serialize() }
+    val name = kind.name
+    storage.updateTitle(groupID, name)
+
+    storage.insertIncomingInfoMessage(context, senderPublicKey, groupID, SignalServiceProtos.GroupContext.Type.UPDATE, SignalServiceGroup.Type.UPDATE, name, members, admins)
+}
+
+private fun MessageReceiver.handleClosedGroupMembersAdded(message: ClosedGroupControlMessage) {
```

Comment:
If you're the admin you want to send the latest encryption key pair to the added members here